qa/tasks/cephfs/test_quiesce.py

```diff
@@ -660,6 +660,40 @@ def test_quiesce_path_splitauth(self):
         op = self.fs.rank_tell(["quiesce", "path", self.subvolume, '--wait'], rank=0)['op']
         self.assertEqual(op['result'], -1) # EPERM
 
+    def test_quiesce_dir_fragment(self):
+        """
+        That quiesce completes with fragmentation in the background.
+        """
+
+        # the config should cause continuous merge-split wars
+        self.config_set('mds', 'mds_bal_split_size', '1') # split anything larger than one item ....
+        self.config_set('mds', 'mds_bal_merge_size', '2') # and then merge if only one item ]:-}
+        self.config_set('mds', 'mds_bal_split_bits', '2')
+
+        self._configure_subvolume()
+
+        self.mount_a.run_shell_payload("mkdir -p root/sub1")
+        self.mount_a.write_file("root/sub1/file1", "file1")
+        self.mount_a.run_shell_payload("mkdir -p root/sub2")
+        self.mount_a.write_file("root/sub2/file2", "file2")
+
+        self.config_set('mds', 'mds_freeze_delay_ms', '15000') # fragments will spend at least 15 seconds freezing
```

Comment:
Commenting on this before I head to bed for the night: did the procedure I outlined in slack not work? Why? I like this config for debugging but I don't think it should be necessary for this test.

Reply:
Sorry, I'm not sure which procedure you're referring to, would you mind repeating?

In this test, the delay is crucial. It reproduces the situation when a directory stays `freezing` for a long time and it does so in a predictable way. As a result, I can be sure that if quiesce completes faster than a directory would reach the `frozen` state we overcome the issue in the ticket.

Reply:
https://ceph-storage.slack.com/archives/D05LGDB5DQE/p1714661056831499?thread_ts=1714581766.305309&cid=D05LGDB5DQE

Here is that test that works for me:

```
    def test_quiesce_parent_frag(self):
        """
        """

        self._configure_subvolume()

        self.mount_a.run_shell_payload("for i in `seq 1 100`; do mkdir -p foo/$i; done")

        path_parent = self.mount_a.cephfs_mntpt + "/foo"

        path_q1 = path_parent + "/1"
        J = self.fs.rank_tell("quiesce", "path", path_q1)
        log.debug(f"{J}")
        reqid = self._reqid_tostr(J['op']['reqid'])
        self._wait_for_quiesce_complete(reqid)

        self.fs.rank_tell("dirfrag", "split", path_parent, "0/0", "1")

        path_q2 = path_parent + "/2"
        J = self.fs.rank_tell("quiesce", "path", path_q2)
        log.debug(f"{J}")
        reqid = self._reqid_tostr(J['op']['reqid'])
        self._wait_for_quiesce_complete(reqid) # will fail without fix

        # TODO: verify dir is fragmented
```

and it can go in the `TestQuiesce` class because it does not require multiple actives. I believe this simpler, no?

Reply:
I don't find it simpler, TBH, though it's similar. Additionally, my version predictably goes into an endless balancing war of tug with as few as two directories and two files, which I think is a stable approach.

Finally, my current test can be moved to `TestQuiesce` as is, too, it's not dependent on the pinning.

Reply:
I'm puzzled you don't find this simpler. It conceptually is three operations:

- quiesce one dir below a parent
- start fragmenting the parent (will  reliably block because step 1 has auth pins for duration of quiesce)
- quiesce another dir below the parent

Your test relies on delaying freezing, configuration-driven fragmentation, and numerous quiesce ops.

Reply:
@batrick I have combined our approaches. Thanks for the trick with two quiesce ops under the same parent, that's the key. I used my infinite rebalancing approach and your double quiesce approach together in a single test with just two subdirectories, and it works well to reproduce the issue. I will push the new version shortly and let you review